[quant][fx] Add more detailed docs for prepare_fx/prepare_qat_fx/convert_fx

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #83132

Summary:
att

Test Plan:
visual inspection of generated docs page
https://pytorch.org/docs/stable/quantization-support.html

Reviewers:

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
[quant][fx] Add more detailed docs for prepare_fx/prepare_qat_fx/convert_fx

Summary:
att

Test Plan:
visual inspection of generated docs page
https://pytorch.org/docs/stable/quantization-support.html

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

andrewor14 commented on torch/ao/quantization/quantize_fx.py:
```diff
@@ -330,19 +330,103 @@ def prepare_fx(
+
+        # We can customize qconfig_mapping in different ways.
+        # e.g. set the global qconfig, which means we will use the same qconfig for
+        # all operators in the model
```
Maybe mention this is only the default qconfig? Right now it sounds a bit like this will override other settings

jerryzh168 replied:
will add a clarification. this reminds me of making it clearer the precedence for different settings, I'll add that to the doc for QConfigMapping as well

andrewor14 commented on torch/ao/quantization/quantize_fx.py:
```diff
@@ -330,19 +330,103 @@ def prepare_fx(
+        # qconfig_mapping = QConfigMapping().set_module_name("linear", qconfig)
+        # e.g. quantize all nn.Linear modules with a specific qconfig
+        # qconfig_mapping = QConfigMapping().set_object_type(torch.nn.Linear, qconfig)
+        # for a more complete list, please see the docstring for `qconfig_mapping`
```
Should we link directly to the class from here?
```
:class:`torch.ao.quantization.QConfigMapping`
```

andrewor14 commented on torch/ao/quantization/quantize_fx.py:
```diff
@@ -330,19 +330,103 @@ def prepare_fx(
+        # for a more complete list, please see the docstring for `qconfig_mapping`
+        # argument
+
+        # example inputs is a tuple of inputs, that is used to infer the type of the
```
nit: example_inputs

andrewor14 replied:
(same in L523)

andrewor14 commented on torch/ao/quantization/quantize_fx.py:
```diff
@@ -330,19 +330,103 @@ def prepare_fx(
+        # TODO: add backend_config after we split the backend_config for fbgemm and qnnpack
+        # e.g. backend_config = get_default_backend_config("fbgemm")
+        # `prepare_fx` inserts observers in the model based on qconfig_mapping and
+        # backend_config, if the configuration for an operator in qconfig_mapping
```
nit: sentence run on, should be
```
backend_config. If the configuration ...
```

andrewor14 commented on torch/ao/quantization/quantize_fx.py:
```diff
@@ -330,19 +330,103 @@ def prepare_fx(
+        #     observer=MinMaxObserver.with_args(dtype=torch.quint8),
+        #     weight=MinMaxObserver.with-args(dtype=torch.qint8))
+        # Note: current qconfig api does not support setting output observer, but
+        # we may have extend this to support these more fine grained control in the
```
Typo? "we may have extend"

jerryzh168 replied:
will remove `have`, probably trying to phrase this in a different way before and forgot to delete `have` here

andrewor14 commented on torch/ao/quantization/quantize_fx.py:
```diff
@@ -330,19 +330,103 @@ def prepare_fx(
+        #    .add_dtype_config(weighted_int8_dtype_config) \
+        #    ...
+
+        # backend_config = BackendConfig().set_backend_pattern_config(linear_pattern_config)
```
Do you think we should leave out the `BackendConfig` example here and only include `backend_config = get_default_backend_config("fbgemm")`? I feel this is a bit long now and most users will probably just use the default.

jerryzh168 replied:
this is a brief documentation for backend_config I think, just like the one for qconfig_mapping, this is an important argument, I think we should explain it briefly at least at the top level.

andrewor14 commented on torch/ao/quantization/quantize_fx.py:
```diff
@@ -382,19 +466,76 @@ def prepare_qat_fx(
+                x = self.linear(x)
+                x = self.sub(x) + x
+                return x
+
```
nit: extra line

andrewor14 commented on torch/ao/quantization/quantize_fx.py:
```diff
@@ -544,7 +696,16 @@ def convert_to_reference_fx(
+        # for more information about reference quantized model
+        # The reference quantized model is a standard format for a quantized model provided
+        # by fx graph mode quantization, it can be further lowered to run on the target
+        # hardware
```
Should we move this paragraph to the main docstring instead of under `Example::`?

jerryzh168 replied:
makes sense, will merge with the main docstring

andrewor14 reviewed (approved):
Looks great! Just some minor comments.

jerryzh168 committed:
Update on "[quant][fx] Add more detailed docs for prepare_fx/prepare_qat_fx/convert_fx"


Summary:
att

Test Plan:
visual inspection of generated docs page
https://pytorch.org/docs/stable/quantization-support.html

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx] Add more detailed docs for prepare_fx/prepare_qat_fx/convert_fx"


Summary:
att

Test Plan:
visual inspection of generated docs page
https://pytorch.org/docs/stable/quantization-support.html

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@pytorchbot merge
